### PR TITLE
noVNC: fix JP keyboard on vmware7+ which uses websocket URL

### LINKED ISSUE #7335: keyboard setting in noVNC not functional when using vmware7
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Console/noVNC
~~~

##### CLOUDSTACK VERSION

~~~
4.17
4.18
~~~

##### SUMMARY

When using vmware 7 and noVNC japanese (and probably all non US) keyboards do not work.

##### STEPS TO REPRODUCE

~~~
log in with language set to a non US/english setting (i.e. japanese)
open a console
~~~

##### EXPECTED RESULTS

~~~
keyboard takes locale input
~~~

##### ACTUAL RESULTS

~~~
keyboard behaves as US/intl
~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes the type error on novnc console of VM on VMware7 which uses Japanese layout.

The other languages layout can be fixed by similar solution.

This PR fixes and closes #7335 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Review by rohityadavcloud — APPROVED** [review]
Change LGTM, didn't test them. I like the overall approach to get/pass the language details from the User VM and have noVNC codebase interpret the keyboard per the passed lang.

However, we need to be careful the next time we upgrade noVNC (to not revert such changes to novnc). Does it make sense to have a novnc fork somewhere (similar to cloudstack-nonoss/noredist repo)?

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: systemvm/agent/noVNC/core/rfb.js [hunk 1]
```diff
@@ -2608,6 +2656,7 @@ export default class RFB extends EventTargetMixin {
                 return true;
 
             case encodings.pseudoEncodingVMwareCursor:
+                this._VMwareCursorSupported = true;
```
Comment (nvazquez) [inline systemvm/agent/noVNC/core/rfb.js:2659]: Is this encoding a requirement to display the JP keyboard?

File: systemvm/agent/noVNC/core/rfb.js [hunk 2]
```diff
@@ -506,13 +521,46 @@ export default class RFB extends EventTargetMixin {
...
+        }
+
         if (this._qemuExtKeyEventSupported && scancode) {
             // 0 is NoSymbol
             keysym = keysym || 0;
 
             Log.Info("Sending key (" + (down ? "down" : "up") + "): keysym " + keysym + ", scancode " + scancode);
 
             RFB.messages.QEMUExtendedKeyEvent(this._sock, keysym, down, scancode);
+        } else if (this._VMwareCursorSupported && Object.keys(this._scancodes).length > 0) {
```
Comment (nvazquez) [inline systemvm/agent/noVNC/core/rfb.js:542]: Could it be the case in which the JP keyboard is needed but uses a different encoding?

File: systemvm/agent/noVNC/core/rfb.js [hunk 3]
```diff
@@ -506,13 +521,46 @@ export default class RFB extends EventTargetMixin {
...
+                    RFB.messages.keyEvent(this._sock, this._shiftKey, 1);
+                }
+                if (! shifted && this._shiftPressed && ! isLetter) {
+                    RFB.messages.keyEvent(this._sock, this._shiftKey, 0);
+                }
+                RFB.messages.VMwareExtendedKeyEvent(this._sock, keysym, down, vscancode_int);
+                if (shifted && ! this._shiftPressed && ! isLetter) {
+                    RFB.messages.keyEvent(this._sock, this._shiftKey, 0);
+                }
+                if (! shifted && this._shiftPressed && ! isLetter) {
```
Comment (nvazquez) [inline systemvm/agent/noVNC/core/rfb.js:558]: Why is this logic repetead after the `VMwareExtendedKeyEvent` call?

File: systemvm/agent/noVNC/core/rfb.js [hunk 4]
```diff
@@ -506,13 +521,46 @@ export default class RFB extends EventTargetMixin {
 
         const scancode = XtScancode[code];
 
+        if (keysym === KeyTable.XK_Shift_L || keysym === KeyTable.XK_Shift_R) {
+            this._shiftPressed = down;
+            this._shiftKey = down ? keysym : KeyTable.XK_Shift_L;
+        }
+        if (keysym === KeyTable.XK_Control_L || keysym === KeyTable.XK_Control_R) {
```
Comment (nvazquez) [inline systemvm/agent/noVNC/core/rfb.js:528]: Seems like ctrlPressed and altPressed are not used, can we remove these 2 checks?

Comment (weizhouapache) [reply systemvm/agent/noVNC/core/rfb.js:528]: @nvazquez yes, I will remove them

File: systemvm/agent/noVNC/core/rfb.js (hunk 3 above)
Comment (weizhouapache) [reply systemvm/agent/noVNC/core/rfb.js:558]: some chars in Japanese keyboard require shift to be pressed.
but if we use different keyboard (e.g. us/uk), we do not need to press shift. in this case, if the vm use Japanese keyboard, we need to press shift at first, send the keycode and release shift.
(this does not apply on letters)

File: systemvm/agent/noVNC/core/rfb.js (hunk 1 above)
Comment (weizhouapache) [reply systemvm/agent/noVNC/core/rfb.js:2659]: this is just to make sure that vm is running on vmware env.
it is probably unnecessary, as we use  `requiresVncOverWebSocketConnection` in java class

File: systemvm/agent/noVNC/core/rfb.js (hunk 2 above)
Comment (weizhouapache) [reply systemvm/agent/noVNC/core/rfb.js:542]: From I have tested, the flag is always true. I will remove it to ensure the console works if different encoding is used

**Comment by weizhouapache** [comment]
@blueorangutan package

File: systemvm/agent/noVNC/core/rfb.js (hunk 3 above)
Comment (weizhouapache) [reply systemvm/agent/noVNC/core/rfb.js:549]: @nvazquez 
From what I have tested, if the vm layout and client keyboard layout are same, this is not required (for example both are Japanese)
However, if the vm layout and client keyboard layout are different, this is required (for example, vm layout is Japanese, but our laptop/client uses US keyboard)

Comment (nvazquez) [reply systemvm/agent/noVNC/core/rfb.js:549]: Ok thanks @weizhouapache +1 for keeping it then

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 vmware-70u3

**Review by andrijapanicsb — APPROVED** [review]
LGTM after testing.

The only difference is when pressing shift + 0 (shift and zero):
- in vCenter is sends key    ~ (even though my onscreen keyboard layout specifies NO key)
- in noVNC console - it doesn't send any key (as expected by my keyboard layout)

Layout below:
image

**Comment by weizhouapache** [comment]
thanks @andrijapanicsb for testing
let's merge this PR and wait for Japanese users to confirm. If it is a bug, we will fix it later (could be a very simple fix)

**Pull request merged**